pypeit/core/datacube.py

```diff
@@ -989,6 +995,339 @@ def compute_weights(all_ra, all_dec, all_wave, all_sci, all_ivar, all_idx, white
+#         detgeom = shapely.strtree.STRtree(detpix_polys)
+#         # Loop through all voxels for this slice and calculate the overlapping area
+#         for wv in range(nvox_wave):
+#             for sp in range(nvox_spat):
+#                 # Generate the voxel coordinates in detector pixel space (points must be counter-clockwise)
+#                 voxel_geom = shapely.geometry.Polygon([(crd_det_spat[wv, sp],   crd_det_spec[wv,   sp]),
+#                                                        (crd_det_spat[wv, sp+1], crd_det_spec[wv,   sp]),
+#                                                        (crd_det_spat[wv, sp+1], crd_det_spec[wv+1, sp]),
+#                                                        (crd_det_spat[wv, sp],   crd_det_spec[wv+1, sp]),
+#                                                        (crd_det_spat[wv, sp],   crd_det_spec[wv,   sp])])
+#                 # Find overlapping detector pixels
+#                 result = detgeom.query(voxel_geom)
+#                 # Sum all overlapping flux-weighted areas
+#                 this_flx = 0
+#                 this_var = 0
+#                 this_area = 0
+#                 for pp in range(len(result)):
+#                     area = voxel_geom.intersection(result[pp]).area
+#                     pix_spat = int(min(result[pp].exterior.coords[0][0], result[pp].exterior.coords[2][0]))
+#                     pix_spec = int(min(result[pp].exterior.coords[0][1], result[pp].exterior.coords[2][1]))
+#                     if ivarimg[pix_spec, pix_spat] != 0.0:
+#                         this_flx += area * fluximg[pix_spec, pix_spat]
+#                         this_var += area**2 * varimgsq[pix_spec, pix_spat]
+#                         this_area += area
+#                 # Fill in the datacube
+#                 this_area = 1 if this_area == 0 else this_area
+#                 datcube[sl, sp, wv] = this_flx / this_area
+#                 varcube[sl, sp, wv] = this_var / this_area**2
+#
+#     # Generate a header
+#     hdr = output_wcs.to_header()
+#
+#     # Add the unit of flux to the header
+#     if fluxcal:
+#         hdr['FLUXUNIT'] = (PYPEIT_FLUX_SCALE, "Flux units -- erg/s/cm^2/Angstrom/arcsec^2")
+#     else:
+#         hdr['FLUXUNIT'] = (1, "Flux units -- counts/s/Angstrom/arcsec^2")
+#
+#     # Save the final datacube
+#     msgs.info("Saving datacube as: {0:s}".format(outfile))
+#     final_cube = DataCube(datcube.T, varcube.T, specname, blaze_wave, blaze_spec, sensfunc=sensfunc, fluxed=fluxcal)
+#     final_cube.to_file(outfile, hdr=hdr, overwrite=overwrite)
+
+
+def generate_cube_subsample(outfile, output_wcs, all_sci, all_ivar, all_wghts, all_wave, tilts, slits, slitid_img_gpm,
+                            astrom_trans, bins, subsample=10, overwrite=False, blaze_wave=None, blaze_spec=None,
+                            fluxcal=False, sensfunc=None, specname="PYP_SPEC", debug=False):
+    """
+    Save a datacube using the subsample algorithm. This algorithm splits
```

Comment:
Maybe add to the docs that the if subsample > 10 the errors will be correlated and the covariance is not being tracked, so the errors won't be right. That is fine, tracking the covariance would be massively painful and this is the tradeoff one has to make between sampling and better looking cubes, and poor sampling and better behaved errors. I would just add a note. Finally, in coadd2d, we called subsample spec_fact and spat_fact. And I think that definition is the inverse of what you have here. It might be best to try to adopt consistent definitions. In other words, spat_fact = 2.0 performs a coarse rebidding, whereas spat_fact = 0.5 would be your subsample = 2. Finally, shouldn't there be an option here to have a different sampling in the spectral direction (is that your z) as opposed to the spatial directions (x,y)?

Reply:
Okay, I see your point about this being different since it has to be an integer. In that case, casting an integer to 1/integer is a bad idea, so let's leave your definitions and variables as they are. So shouldn't this be called supersample though? You are not subsampling if your final cube has higher resolution than your individual frames, you are super-sampling right?